lib/sidekiq/pool/cli.rb

```diff
@@ -228,25 +248,43 @@ def handle_master_signal(sig)
           logger.info "Sending #{sig} signal to the pool"
           signal_to_pool(sig)
         when 'HUP'
-          logger.info 'Gracefully reloading pool'
-          old_pool = @pool.dup
+          reload
+        end
+      end
+
+      def reload
+        logger.info 'Gracefully reloading pool'
 
-          # Signal old pool
-          # USR1 tells Sidekiq it will be shutting down in near future.
-          signal_to_pool('USR1') if @suspend_before_graceful_reload
+        # USR1 tells Sidekiq it will be shutting down in near future.
+        signal_to_pool('USR1') if @suspend_before_graceful_reload
 
-          # Reset pool
-          @pool = []
+        add_reload_marker
+        reexec
 
-          # Start new pool
-          start_new_pool
+        stop_children
+        logger.info 'Graceful reload completed'
+      ensure
+        exit(0)
+      end
 
-          # Stop old pool
-          stop_children(old_pool)
-          logger.info 'Graceful reload completed'
+      def reexec
+        fork do
+          cmd = [START_CTX[0]].concat(START_CTX[:argv])
+          logger.info("Starting new master process #{cmd}")
+          exec(*cmd)
```

Comment:
Will new process retain `ENV` from old process?

Reply:
@laurynas yes, environment is inherited